aquifer/aquifer
# Sync API

Abstracts syncing operations in build to a new Sync API. It has become clear to me that it'll be useful to facilitate syncing of items in more places that just after the build is scaffolded, like pre-build for example. This API creates a standardized way of initiating syncing of items.

This refactors a significant portion of the build process. Pay special attention to my somewhat lacking understanding of Promises and my code using them herein 😬 .

**To test:**
- [x] Run `aquifer create test -d 8 && cd test`
- [x] Run `aquifer build`
- [x] Verify that the `/build` directory contains a Drupal site and the synced directories and files defined in `aquifer.json` are synced appropriately.
- [x] Run `mkdir -p config/base`
- [x] Add the following to the `sync.directories` object in `aquifer.json`:

```
"config": {
  "destination": "sites/default/config",
  "method": "symlink",
  "conflict": "overwrite",
  "type": "dir",
  "hook": "preBuild"
}
```
- [x] Run `aquifer build` 
- [x] Verify console output shows this getting synced during the preBuild sync operation.
- [x] Verify the new `/config` directory is symlinked into `/build/sites/default`.
- [x] Run `touch root/LICENSE.txt`
- [x] Add the following to the `sync.files` object in `aquifer.json`:

```
"root/LICENSE.txt": {
  "destination": "LICENSE.txt",
  "method": "copy",
  "conflict": "overwrite",
  "type": "file",
  "hook": "postBuild"
}
```
- [x] Run `aquifer build`
- [x] Verify console output shows this getting synced during the postBuild sync operation.
- [x] Verify `/build/LICENSE.txt` is overwritten by a copy (not symlink) or your blank file.
- [x] In `aquifer.json` change the conflict value to skip for `root/LICENSE.txt`.
- [x] Run `aquifer build` and verify `/build/LICENSE.txt` is has been left alone.

We also need to verify this is working with aquifer-git's `addLinks` and `excludeLinks` options so it would be great if someone else can test this that has `aquifer-git` configured for a site:
- [x] Install `aquifer-git` in your project ensuring it is on the `1.0.0` branch.
- [x] Edit your `aquifer.json` file to include `aquifer-git` configuration:
  
  ```
  "extensions": {
    "aquifer-git": {
      "source": "aquifer-git",
      "remote": "user@agitrepositoryhost.com:repositoryname.git",
      "branch": "master",
      "folder": "docroot",
      "name": "Deploy Bot",
      "email": "deploybot@aquifer.io",
      "excludeLinks": ["sites/default/files"],
      "addLinks": [
        {
          "src": "pizza",
          "dest": "pizza",
          "type": "dir"
        }
      ],
      "delPatterns": ["*", "!.git"]
    }
  }
  ```
- [x] Add a file to the files directory:
  
  ```
  touch files/beer.js
  ```
- [x] Add the pizza directory you just defined to the project root:
  
  ```
  mkdir pizza && touch pizza/pepperoni.txt
  ```
- [x] Run `aquifer deploy-git --debug`
- [x] Inspect the `aquifer-git-XXXXXXX` directory that is left behind.
- [x] Verify that `sites/default/files` does not exist or if it does that it has not been copied (i.e. does not contain `beer.js`)
- [x] Verify that the `pizza` directory has been copied.

**Test "contents" property**
- [x] Add a new directory and files that will copy the source contents only:
  
  ```
  mkdir beer && touch beer/ipa.txt && touch beer/stout.txt
  ```
- [x] Add the directory to the sync config in `aquifer.json` with the "contents" property set to true:

```
"beer": {
  "destination": "sites/default",
  "method": "symlink",
  "conflict": "overwrite",
  "type": "dir",
  "hook": "postBuild",
  "contents": true
}
```
- [x] Also add the "contents" property to the existing "root" sync item (or the build will fail):

```
"root": {
  "destination": "",
  "method": "symlink",
  "conflict": "overwrite",
  "type": "dir",
  "hook": "postBuild",
  "required": true,
  "contents": true
},
```
- [x] Run `aquifer build`
- [x] Verify that `ipa.txt` and `stout.txt` were copied into `/sites/default`.
